common/src/main/java/io/pravega/common/util/ResourcePool.java

```diff
@@ -0,0 +1,286 @@
+ * If a returned resource is invalid, it is destroyed and a replacement resource is created to fulfill waiting requests. 
+ * If there are no waiting requests, this class tries to maintain an idle pool of resources of maxidleSize. 
+ * Once maximum allowed idle resources are present with it, any additional resource returned to it is discarded.    
+ * 
+ * The users of this class are expected to return the used resource back to the pool exactly once. The behaviour is 
+ * nondeterministic if returned multiple times.  
+ * 
+ * The ResourcePool can be shutdown. However, the shutdown trigger does not prevent callers to attempt 
+ * to create new resources as long as they have a valid handle to the resource pool. 
+ * Shutdown ensures that it drains all idle resources once shutdown is triggered.  
+ */
+public class ResourcePool<T> {
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private final ArrayDeque<T> idleResources;
+    @GuardedBy("lock")
+    private boolean isRunning;
+    @GuardedBy("lock")
+    private final ArrayDeque<WaitingRequest<T>> waitQueue;
+    @GuardedBy("lock")
+    private int resourceCount;
+    private final int maxConcurrent;
+    private final int maxIdle;
+
+    private final Listener listener;
+    private final Supplier<CompletableFuture<T>> tSupplier;
+    private final Consumer<T> tDestroyer;
+
+    public ResourcePool(Supplier<CompletableFuture<T>> tSupplier, Consumer<T> tDestroyer, int maxConcurrent, int maxIdle) {
+        this(tSupplier, tDestroyer, maxConcurrent, maxIdle, null);
+    }
+
+    @VisibleForTesting
+    ResourcePool(Supplier<CompletableFuture<T>> tSupplier, Consumer<T> tDestroyer,
+                 int maxConcurrent, int maxIdle, Listener listener) {
+        Preconditions.checkNotNull(tSupplier);
+        Preconditions.checkNotNull(tDestroyer);
+        Preconditions.checkArgument(maxConcurrent >= maxIdle);
+        Preconditions.checkArgument(maxIdle >= 0);
+        this.idleResources = new ArrayDeque<>();
+        this.isRunning = true;
+        this.waitQueue = new ArrayDeque<>();
+        this.resourceCount = 0;
+        this.maxConcurrent = maxConcurrent;
+        this.maxIdle = maxIdle;
+        this.listener = listener;
+        this.tSupplier = tSupplier;
+        this.tDestroyer = tDestroyer;
+    }
+
+    /**
+     * Method to get a resource initialized with supplied arg.
+     * This method attempts to find an existing available resource.
+     * If not found, it submits a new waiting request for whenever a resource becomes available. A resource could become available
+     * because such a resource was returned to the pool or a new resource was created.
+     * It also opportunistically submits a request to create a new resource if required.
+     *
+     * @return A completableFuture which when completed will have the resource object that the caller requested.
+     */
+    public CompletableFuture<T> getResource() {
```

Comment:
This looks good. However I have a suggestion that may improve the usability of this. 

Currently we need to call `getResource()` and then `returnResource` when we're done with a resource. This will force the caller to hold a reference to this pool, which, while doable, may not produce the nicest code.

One thing I've done in the past is to define a nested class `AutoReturnResource<T> implements AutoCloseable`. This can only be created in the `ResourcePool` class and will be returned by `getResource`. This new class will have a getter providing access to the resource, but upon calling `close` it will internally invoke the pool's `returnResource` (which will now be made private).

Both implementations are equivalent, but IMO the `AutoReturnResource` looks a bit more elegant and makes it obvious to the user that it's a finite resource and needs to be cleaned up because it implements `AutoCloseable`.